### PR TITLE
Unsafe Getters:  Alternative to PR #103   which fixes issue #102

### PR DESCRIPTION
**Pull request by ndinsmore**
This is an alternative to PR #103 by @odow  that doesn't cache the values but instead uses an unsafe getter function to the CBC interface only where it is clearly safe.   Does not need to copy entire array.

### CODE REVIEW

**Comment by blegat** [comment]
Could you post the results of this benchmark: https://github.com/JuliaOpt/Cbc.jl/issues/102#issue-406542463 before and after your changes ?

**Comment by ndinsmore** [comment]
master:
```
BenchmarkTools.Trial: 
  memory estimate:  6.72 GiB
  allocs estimate:  509495
  --------------
  minimum time:     6.101 s (16.10% GC)
  median time:      6.101 s (16.10% GC)
  mean time:        6.101 s (16.10% GC)
  maximum time:     6.101 s (16.10% GC)
  --------------
  samples:          1
  evals/sample:     1
```

ndinsmore:unsafe_getters
```
BenchmarkTools.Trial: 
  memory estimate:  9.38 MiB
  allocs estimate:  449495
  --------------
  minimum time:     27.054 ms (0.00% GC)
  median time:      36.035 ms (0.00% GC)
  mean time:        39.875 ms (5.44% GC)
  maximum time:     143.152 ms (0.00% GC)
  --------------
  samples:          126
  evals/sample:     1
```
od/fix_variable_query
```
BenchmarkTools.Trial: 
  memory estimate:  6.63 MiB
  allocs estimate:  389495
  --------------
  minimum time:     26.278 ms (0.00% GC)
  median time:      37.727 ms (0.00% GC)
  mean time:        39.522 ms (6.03% GC)
  maximum time:     72.969 ms (22.17% GC)
  --------------
  samples:          127
  evals/sample:     1
```

**Comment by odow** [comment]
No worries. Since the times are the same, my vote is to reduce the complexity of the C wrapper as much as possible and keep any MOI-specific complexity in the MOI wrapper.

Thoughts?

**Comment by blegat** [comment]
Why is it MOI-specific ? If the Cbc API provides an unsafe methods that does a copy, it makes sense to mirror it in the Julia wrapper, even more if it turns out the be useful for the MOI wrapper.

File: src/CbcCInterface.jl [hunk 1]
```diff
@@ -214,21 +214,25 @@ end
 @getproperty Float64 getObjSense
 
 for s in (:getRowLower, :getRowUpper, :getRowActivity)
-    @eval function ($s)(prob::CbcModel)
+    unsafe_s = Symbol("unsafe", '_', s)
+    @eval function ($unsafe_s)(prob::CbcModel)
```
Comment (mlubin) [inline src/CbcCInterface.jl]: Please add comments explaining why this is unsafe (i.e., It returns an array whose memory is owned by Cbc. The underlying memory could be invalidated whenever Cbc feels like it.)

File: src/CbcCInterface.jl [hunk 2]
```diff
@@ -214,21 +214,25 @@ end
...
         check_problem(prob)
         nrow = Int(getNumRows(prob))
         p = @cbc_ccall $s Ptr{Float64} (Ptr{Cvoid},) prob
-        return copy(unsafe_wrap(Array,p,(nrow,)))
+        return unsafe_wrap(Array,p,(nrow,))
```
Comment (mlubin) [inline src/CbcCInterface.jl]: Nit: spaces after comments to follow the JuMP style guide: `unsafe_wrap(Array, p, (nrow,))`. Yes, this was already there but when touching a line you should generally fix style issue.

File: src/CbcCInterface.jl [hunk 3]
```diff
@@ -214,21 +214,25 @@ end
...
     end
+    @eval ($s)(prob::CbcModel)=copy(($unsafe_s)(prob))
```
Comment (mlubin) [inline src/CbcCInterface.jl]: Style: Missing space before and after `=`.

**Comment by ndinsmore** [comment]
Everything should be ready to go

File: src/CbcCInterface.jl [hunk 4]
```diff
@@ -57,6 +57,15 @@ export CbcModel,
     secondaryStatus,
     addSOS
 
+#The follow unsafe methods are created but not exported
```
Comment (mlubin) [inline src/CbcCInterface.jl]: We don't list unexported functions, so it's inconsistent to list these here. Please remove.
See also the new guidance on naming unexported functions: http://www.juliaopt.org/JuMP.jl/dev/style/#Exported-and-non-exported-names-1. That's not necessary to address right now.

File: src/CbcCInterface.jl [hunk 5]
```diff
@@ -214,21 +223,33 @@ end
 @getproperty Float64 getObjSense
 
 for s in (:getRowLower, :getRowUpper, :getRowActivity)
-    @eval function ($s)(prob::CbcModel)
+    #This macro creates a unsafe wrapper to the following function
```
Comment (mlubin) [inline src/CbcCInterface.jl]: Misting space after `#` and punctuation at the end of the sentences (http://www.juliaopt.org/JuMP.jl/dev/style/#Comments-1).

File: src/CbcCInterface.jl [hunk 6]
```diff
@@ -214,21 +223,33 @@ end
...
+    #The unsafe methods should be recalled when ever a change has been made to Cbc
+    unsafe_s = Symbol("unsafe", '_', s)
+    @eval function ($unsafe_s)(prob::CbcModel)
         check_problem(prob)
         nrow = Int(getNumRows(prob))
         p = @cbc_ccall $s Ptr{Float64} (Ptr{Cvoid},) prob
-        return copy(unsafe_wrap(Array,p,(nrow,)))
+        return unsafe_wrap(Array, p, (nrow,))
     end
+    #By making a copy of the unsafe array we can depend on it now changing.
```
Comment (odow) [inline src/CbcCInterface.jl]: we can depend on it *not* changing?

File: src/CbcCInterface.jl [hunk 7]
```diff
@@ -214,21 +223,33 @@ end
...
+    #The unsafe methods should be recalled when ever a change has been made to Cbc
+    unsafe_s = Symbol("unsafe", '_', s)
+    @eval function ($unsafe_s)(prob::CbcModel)
         check_problem(prob)
         ncol = Int(getNumCols(prob))
         p = @cbc_ccall $s Ptr{Float64} (Ptr{Cvoid},) prob
-        return copy(unsafe_wrap(Array,p,(ncol,)))
+        return unsafe_wrap(Array, p, (ncol,))
     end
+    #By making a copy of the unsafe array we can depend on it now changing.
```
Comment (odow) [inline src/CbcCInterface.jl]: Ditto not.

File: src/CbcCInterface.jl (hunk 4 above)
Comment (ndinsmore) [reply src/CbcCInterface.jl]: So should I export them or make them underscored?

Comment (mlubin) [reply src/CbcCInterface.jl]: Just removing this comment is fine for now (to make it consistent with the rest of the code). Eventually when this package is updated to follow the new conventions, the functions should have underscores.

Comment (ndinsmore) [reply src/CbcCInterface.jl]: As they are being used outside of the CbcCInterface module I opted to export them

File: src/CbcCInterface.jl [hunk 8]
```diff
@@ -59,6 +66,9 @@ export CbcModel,
 
 
 
+
+
+
```
Comment (mlubin) [inline src/CbcCInterface.jl]: Nit: Please remove these added blank lines.

**Pull request merged**